filecoin-proofs/src/bin/phase2.rs

```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+        prev_param_number,
+        param_number
+    );
+
+    if param_size.is_large() {
+        info!("large param file found, contributing as small");
+    }
+
+    // Get OS entropy prior to deserialization the previous participant's params.
+    let seed = get_mixed_entropy();
+    let mut rng = ChaChaRng::from_seed(seed);
+
+    /*
+    let path_after =
+        params_filename(proof, hasher, sector_size, &head, param_number, param_size);
+    */
+    // Default to small params for first participant.
+    let path_after = params_filename(
         proof,
         hasher,
-        display_sector_size(sector_size),
-        head
+        sector_size,
+        &head,
+        param_number,
+        ParamSize::Small,
+        false,
     );
 
-    let mut next_before_params: Option<MPCParameters> = None;
-    let mut next_before_filename: Option<String> = None;
-    let mut param_number: usize = 0;
-
-    loop {
-        let (before_params, before_filename) = if next_before_params.is_some() {
-            let before_params = next_before_params.take().unwrap();
-            let before_filename = next_before_filename.take().unwrap();
-            (before_params, before_filename)
-        } else {
-            let before_filename = params_filename(proof, hasher, sector_size, &head, param_number);
-            let before_path = Path::new(&before_filename);
-            if !before_path.exists() {
-                info!("waiting for params file: {}", before_filename);
-                while !before_path.exists() {
-                    sleep(Duration::from_secs(SLEEP_SECS));
-                }
-            }
-            info!("found file: {}", before_filename);
-            info!("reading params file: {}", before_filename);
-            let file = File::open(&before_path).unwrap();
-            let mut reader = BufReader::with_capacity(1024 * 1024, file);
-            let read_start = Instant::now();
-            let before_params = MPCParameters::read(&mut reader, true).unwrap();
-            info!(
-                "successfully read params, dt_read={}s",
-                read_start.elapsed().as_secs()
+    let start_total = Instant::now();
+    let start_read = Instant::now();
+
+    let mut small_params = if param_size.is_large() {
+        info!("reading large 'before' params as small: {}", path_before);
+        read_small_params_from_large_file(&path_before).unwrap_or_else(|e| {
+            panic!(
+                "failed to read large param file `{}` as small: {}",
+                path_before, e
             );
-            param_number += 1;
-            (before_params, before_filename)
-        };
+        })
+    } else {
+        info!("reading small 'before' params as small: {}", path_before);
+        let file_before = File::open(path_before).unwrap();
+        let mut reader = BufReader::with_capacity(1024 * 1024, file_before);
+        let read_raw_affine = false; // FIXME: need to support the true path also.
```

Comment:
We can just pass `false` in as an argument.

Reply:
Since it's now okay to create an `MPCSmall` from raw params, I have instead removed this line and pass the value of `read_raw` received from parsing the input file. This resolves the FIXME.

Reply:
Nice, thanks.